core/src/main/java/org/testcontainers/containers/ComposeServiceWaitStrategyTarget.java

```diff
@@ -0,0 +1,68 @@
+package org.testcontainers.containers;
+
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import com.github.dockerjava.api.model.Container;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import org.slf4j.Logger;
+import org.testcontainers.DockerClientFactory;
+import org.testcontainers.containers.wait.strategy.WaitStrategyTarget;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Class to provide a wait strategy target for services started through docker-compose
+ */
+@EqualsAndHashCode
+class ComposeServiceWaitStrategyTarget implements WaitStrategyTarget {
+
+    private final Container container;
+    private final GenericContainer proxyContainer;
+    @Getter
+    private final Logger logger;
+    private Map<Integer, Integer> mappedPorts = new HashMap<>();
+    @Getter
+    private List<Integer> exposedPorts = new ArrayList<>();
+    private InspectContainerResponse containerInfo;
+
+    ComposeServiceWaitStrategyTarget(Container container, GenericContainer proxyContainer,
+                                     Logger logger, Map<Integer, Integer> mappedPorts) {
+        this.container = container;
+
+        this.proxyContainer = proxyContainer;
+        this.logger = logger;
+
+        if (mappedPorts != null) {
+            this.mappedPorts.putAll(mappedPorts);
+            this.exposedPorts.addAll(this.mappedPorts.keySet());
```

Comment:
the `getExposedPorts()` method from the `ContainerState` interface returns `Set<Integer>`, while `this.mappedPorts.keySet()` returns `List<Integer>`. 

I didn't want to create a new `Set` every time the getter is called.

Reply:
`keySet()` returns `Set` as far as I remember :) Maybe you meant that `getExposedPorts` returns `List<Integer>` and not `Set<Integer>`? 

Also, maybe we should make `getExposedPortNumbers` return `Set<Integer>`, use it where possible, and add `getExposedPorts` to `ComposeServiceWaitStrategyTarget` with `getExposedPortNumbers().stream.collect(toList())` implementation?

Reply:
Yes, sorry, obviously `keySet()` returns a `Set`! :-)

> getExposedPortNumbers().stream.collect(toList())

Isn't that functionally the same as `new ArrayList<>(getExposedPortNumbers());`?

Reply:
it is :) Both are equally fine to me since we're talking about collections with just a few elements

Reply:
> add getExposedPorts to ComposeServiceWaitStrategyTarget with getExposedPortNumbers().stream.collect(toList()) implementation?

Ah, I've just realised why this won't work. `getExposedPortNumbers()` in `ContainerState` is a default method that calls `getExposedPort()`!

```java
default Set<Integer> getExposedPortNumbers() {
        return getExposedPorts().stream()
            .map(this::getMappedPort)
            .collect(Collectors.toSet());
    }
```

Reply:
just override (implement) both in`ComposeServiceWaitStrategyTarget` , `getExposedPortNumbers` and `getExposedPorts`

Reply:
Despite the similar names, `getExposedPorts()` and `getExposedPortNumbers()` do different things.

I think `getExposedPortNumbers()` is confusingly named, because what it actually does is return the port on the `SocatContainer` that the service port has been mapped to.

e.g. `.withExposedService("redis_1", 6379)`.
`getExposedPorts()` returns 6379 and `getExposedPortNumbers()` would return e.g. 2000

I took the name from a private method in `GenericContainer`, but it should maybe be renamed in `ContainerState` to `getMappedPorts()`. WDYT?

Reply:
oh, I didn't notice that! Yes, sure!

P.S. I checked the implementation and it seems it is only used in `WaitStrategyTarget ::getLivenessCheckPortNumbers`.

Let's just remove this method and use it's implementation in `WaitStrategyTarget`.
WDYT?

Reply:
Despite the similar names, `getExposedPorts()` and `getExposedPortNumbers()` do different things.

I think `getExposedPortNumbers()` is confusingly named, because what it actually does is return the port  that the service port has been mapped to.

e.g. `.withExposedService("redis_1", 6379)`.
`getExposedPorts()` returns 6379 and `getExposedPortNumbers()` would return e.g. 32679

I took the name from a private method in `GenericContainer`, but it should maybe be renamed in `ContainerState` to `getMappedPorts()`. WDYT?

Reply:
Sorry, I deleted the original comment, because I was starting to confuse myself.
Yes, it's only used in `WaitStrategyTarget ::getLivenessCheckPortNumbers`
I agree, it can definitely go!